subsys/lorawan/lorawan.c

```diff
@@ -0,0 +1,457 @@
+
+	mib_req.Type = MIB_NWK_KEY;
+	mib_req.Param.NwkKey = join_cfg->otaa.nwk_key;
+	LoRaMacMibSetRequestConfirm(&mib_req);
+
+	mib_req.Type = MIB_APP_KEY;
+	mib_req.Param.JoinEui = join_cfg->otaa.app_key;
+	LoRaMacMibSetRequestConfirm(&mib_req);
+
+	return LoRaMacMlmeRequest(&mlme_req);
+}
+
+static LoRaMacStatus_t lorawan_join_abp(
+			const struct lorawan_join_config *join_cfg)
+{
+	MibRequestConfirm_t mib_req;
+
+	mib_req.Type = MIB_ABP_LORAWAN_VERSION;
+	mib_req.Param.AbpLrWanVersion.Value = LORAWAN_ABP_VERSION;
+	LoRaMacMibSetRequestConfirm(&mib_req);
+
+	mib_req.Type = MIB_NET_ID;
+	mib_req.Param.NetID = 0;
+	LoRaMacMibSetRequestConfirm(&mib_req);
+
+	mib_req.Type = MIB_DEV_ADDR;
+	mib_req.Param.DevAddr = join_cfg->abp.dev_addr;
+	LoRaMacMibSetRequestConfirm(&mib_req);
+
+	mib_req.Type = MIB_F_NWK_S_INT_KEY;
+	mib_req.Param.FNwkSIntKey = join_cfg->abp.nwk_skey;
+	LoRaMacMibSetRequestConfirm(&mib_req);
+
+	mib_req.Type = MIB_S_NWK_S_INT_KEY;
+	mib_req.Param.SNwkSIntKey = join_cfg->abp.nwk_skey;
+	LoRaMacMibSetRequestConfirm(&mib_req);
+
+	mib_req.Type = MIB_NWK_S_ENC_KEY;
+	mib_req.Param.NwkSEncKey = join_cfg->abp.nwk_skey;
+	LoRaMacMibSetRequestConfirm(&mib_req);
+
+	mib_req.Type = MIB_APP_S_KEY;
+	mib_req.Param.AppSKey = join_cfg->abp.app_skey;
+	LoRaMacMibSetRequestConfirm(&mib_req);
+
+	mib_req.Type = MIB_NETWORK_ACTIVATION;
+	mib_req.Param.NetworkActivation = ACTIVATION_TYPE_ABP;
+	LoRaMacMibSetRequestConfirm(&mib_req);
+
+	return LORAMAC_STATUS_OK;
+}
+
+int lorawan_join(const struct lorawan_join_config *join_cfg)
+{
+	LoRaMacStatus_t status;
+	int ret = 0;
+
+	k_mutex_lock(&lorawan_join_mutex, K_FOREVER);
+
+	if (join_cfg->mode == LORAWAN_ACT_OTAA) {
+		status = lorawan_join_otaa(join_cfg);
+		if (status != LORAMAC_STATUS_OK) {
+			LOG_ERR("OTAA join failed: %s",
+				lorawan_status2str(status));
+			ret = lorawan_status2errno(status);
+			goto out;
+		}
+
+		LOG_DBG("Network join request sent!");
+
+		/*
+		 * We can be sure that the semaphore will be released for
+		 * both success and failure cases after a specific time period.
+		 * So we can use K_FOREVER and no need to check the return val.
+		 */
+		k_sem_take(&mlme_confirm_sem, K_FOREVER);
+		if (last_mlme_confirm_status != LORAMAC_EVENT_INFO_STATUS_OK) {
+			ret = lorawan_eventinfo2errno(last_mlme_confirm_status);
+			goto out;
+		}
+	} else if (join_cfg->mode == LORAWAN_ACT_ABP) {
+		status = lorawan_join_abp(join_cfg);
+		if (status != LORAMAC_STATUS_OK) {
+			LOG_ERR("ABP join failed: %s",
+				lorawan_status2str(status));
+			ret = lorawan_status2errno(status);
+			goto out;
+		}
+	} else {
+		ret = -EINVAL;
+	}
+
+out:
+	k_mutex_unlock(&lorawan_join_mutex);
+	return ret;
+}
+
+int lorawan_set_class(enum lorawan_class dev_class)
+{
+	LoRaMacStatus_t status;
+	MibRequestConfirm_t mib_req;
+
+	mib_req.Type = MIB_DEVICE_CLASS;
+
+	switch (dev_class) {
+	case LORAWAN_CLASS_A:
+		mib_req.Param.Class = CLASS_A;
+		break;
+	case LORAWAN_CLASS_C:
```

Comment:
class B is not managed, no problem with this, but as it is declared in enum, maybe it should be interesting to put LOG_ERR to inform that it is not developped yet. What do you think about it ?

Reply:
Hmm, makes sense! But I should admit that CLASS_C itself is not supported yet. So I'll remove it too and print error message.

Reply:
I tested class C a long time ago and it was working. I can't remember if I needed to change anything though. Unfortunately, I can't easily test it since I have repurposed my custom gateway (there are 2 TTN gateways nearby anyway) to do other things.